cpp/src/io/parquet/reader_impl_helpers.cpp

```diff
@@ -536,6 +554,304 @@ aggregate_reader_metadata::aggregate_reader_metadata(
+                           thrust::make_counting_iterator(schema_elem.num_children),
+                           [&](auto const& idx) {
+                             return validate_schemas(arrow_schema.children[idx],
+                                                     schema_elem.children_idx[idx]);
+                           });
+      } else {
+        return false;
+      }
+    };
+
+  // Function to co-walk arrow and parquet schemas
+  std::function<void(arrow_schema_data_types&, int)> co_walk_schemas =
+    [&](arrow_schema_data_types& arrow_schema, int schema_idx) {
+      auto& pq_schema_elem = per_file_metadata[0].schema[schema_idx];
+      std::for_each(thrust::make_counting_iterator(0),
+                    thrust::make_counting_iterator(pq_schema_elem.num_children),
+                    [&](auto const& idx) {
+                      co_walk_schemas(arrow_schema.children[idx], pq_schema_elem.children_idx[idx]);
+                    });
+
+      // true for DurationType columns only for now.
+      if (arrow_schema.type.id() != type_id::EMPTY) {
+        pq_schema_elem.arrow_type = arrow_schema.type.id();
+      }
+    };
+
+  // Get Parquet schema root
+  auto pq_schema_root = get_schema(0);
+
+  // verify equal number of children for both schemas at root level
+  if (pq_schema_root.num_children != static_cast<int32_t>(arrow_schema_root.children.size())) {
+    CUDF_LOG_DEBUG("Parquet reader encountered a mismatch between Parquet and arrow schema.",
+                   "arrow:schema not processed.");
+    return;
+  }
+
+  // Verify equal number of children at all sub-levels
+  if (not std::all_of(thrust::make_counting_iterator(0),
+                      thrust::make_counting_iterator(pq_schema_root.num_children),
+                      [&](auto const& idx) {
+                        return validate_schemas(arrow_schema_root.children[idx],
+                                                pq_schema_root.children_idx[idx]);
+                      })) {
+    CUDF_LOG_DEBUG("Parquet reader encountered a mismatch between Parquet and arrow schema.",
+                   "arrow:schema not processed.");
+    return;
+  }
+
+  // All good, now co-walk schemas
+  std::for_each(thrust::make_counting_iterator(0),
+                thrust::make_counting_iterator(pq_schema_root.num_children),
+                [&](auto const& idx) {
+                  co_walk_schemas(arrow_schema_root.children[idx],
+                                  pq_schema_root.children_idx[idx]);
+                });
+}
+
+std::optional<std::string_view> aggregate_reader_metadata::decode_ipc_message(
+  std::string_view const serialized_message) const
+{
+  // Constants copied from arrow source and renamed to match the case
+  constexpr auto MESSAGE_DECODER_NEXT_REQUIRED_SIZE_INITIAL         = sizeof(int32_t);
+  constexpr auto message_decoder_next_required_size_metadata_length = sizeof(int32_t);
+  constexpr int32_t IPC_CONTINUATION_TOKEN                          = -1;
```

Comment:
I'm curious as to why one of these is lowercase

Reply:
I almost vividly remember changing all these to uppercase but oh well.